### PR TITLE
Feature/che 281 refactor architecture

### PR DESCRIPTION
**Pull request by EvansPie**
CHE-281

# What this PR does

Refactor Primer SDK to the clean arch.

# Notable decisions & other stuff

I checked the following:

- [x] Network requests only called from Services
- [x] Services only called from View Models
- [x] View Models are registered as dependencies and are only called from View Controllers (with an exception for the `FormView` that will be fixed when it gets refactored)
- [x] Views are only created within view controllers
- [x] Helpers have been moved outside of classes into the Extensions & Utilities folder

# Before merging

_QA_

- [x] I manually tested this with a demo application (simulator)
- [ ] I manually tested this with a demo application (real device)
- [ ] I wrote unit tests for each new util-like function
- [ ] I wrote at least basic e2e tests for the new functionalities or new paths
- [ ] If this PR introduces UI changes, I manually tested this PR on iPhone 8 (iOS 10, 12, 14), iPhone 12, and iPhone 12 Max
- [ ] If this PR modifies the behavior of an API, I did my best to make my new API backward compatible
- [ ] If this PR modifies the API, I attempted to implement the feature with an example

_Documentation_

- [ ] If this PR modifies the API, I updated the API Reference.
- [ ] If this PR adds new options to the API, I wrote some documentation or guide in the online documentation
- [ ] If the PR modifies the API, I wrote some documentation to deprecate it

# After merging

- Make sure a new release has been pushed to Cocoapods trunk

### CODE REVIEW

File: PrimerSDK/Classes/Data Models/Consolable.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+//  Consolable.swift
+//  PrimerSDK
+//
+//  Created by Evangelos Pittas on 19/3/21.
+//
+
+import Foundation
+
+@propertyWrapper
+struct Consolable<T> {
```
Comment (ceriksson) [inline PrimerSDK/Classes/Data Models/Consolable.swift:11]: let's discuss this one when you got time  :)

File: PrimerSDK/Classes/Extensions & Utilities/BundleExtension.swift [hunk 2]
```diff
@@ -0,0 +1,20 @@
+//
+//  BundleExtension.swift
+//  PrimerSDK
+//
+//  Created by Evangelos Pittas on 18/3/21.
+//
+
+import Foundation
+
+internal extension Bundle {
```
Comment (ceriksson) [inline PrimerSDK/Classes/Extensions & Utilities/BundleExtension.swift:10]: let's also discuss this.

File: PrimerSDK/Classes/Extensions & Utilities/StringExtension.swift [hunk 3]
```diff
@@ -98,6 +127,16 @@ extension String {
         return mutableAttString
     }
     
+    func localized(comment: String? = nil) -> String {
```
Comment (ceriksson) [inline PrimerSDK/Classes/Extensions & Utilities/StringExtension.swift:130]: we should probably also have a quick chat about localization.

File: PrimerSDK/Classes/Extensions & Utilities/UIDeviceExtension.swift [hunk 4]
```diff
@@ -0,0 +1,90 @@
+//
+//  UIDeviceExtension.swift
+//  PrimerSDK
+//
+//  Created by Evangelos Pittas on 18/3/21.
+//
+
+import UIKit
+
+internal extension UIDevice {
```
Comment (ceriksson) [inline PrimerSDK/Classes/Extensions & Utilities/UIDeviceExtension.swift:10]: what does this util do?

File: PrimerSDK/Classes/User Interface/PCI/CardForm/CardFormViewModel.swift [hunk 5]
```diff
@@ -31,6 +31,7 @@ class CardFormViewModel: CardFormViewModelProtocol {
         }
     }
     
+    // FIXME: Is this called only by the unit tests?
     func tokenize(instrument: PaymentInstrument, completion: @escaping (Error?) -> Void) {
```
Comment (ceriksson) [inline PrimerSDK/Classes/User Interface/PCI/CardForm/CardFormViewModel.swift:35]: This entire view & view model should be deleted and replaced by FormView (this has not been done out of caution but should be fine)

File: PrimerSDK/Classes/User Interface/Primer/MultiCardIconComponent.swift [hunk 6]
```diff
@@ -7,6 +7,7 @@
 
 import UIKit
 
+// FIXME: Unused?
```
Comment (ceriksson) [inline PrimerSDK/Classes/User Interface/Primer/MultiCardIconComponent.swift:10]: Most likely unused, did you comment it out and it was fine? (if so let's remove)

File: PrimerSDK/Classes/User Interface/Primer/PaymentMethodComponent.swift [hunk 7]
```diff
@@ -7,6 +7,7 @@
 
 import UIKit
 
+// FIXME: Unused?
 class PaymentMethodComponent: UIView {
```
Comment (ceriksson) [inline PrimerSDK/Classes/User Interface/Primer/PaymentMethodComponent.swift:11]: if not used anywhere at the moment, let's remove

File: PrimerSDK/Classes/User Interface/Primer/SingleCardIconComponent.swift [hunk 8]
```diff
@@ -7,6 +7,7 @@
 
 import UIKit
 
+// FIXME: Unused?
 class SingleCardIconComponent: UIView {
```
Comment (ceriksson) [inline PrimerSDK/Classes/User Interface/Primer/SingleCardIconComponent.swift:11]: same for this (remove if lint says not used).

**Review by ceriksson — APPROVED** [review]
lgtm

Let's schedule a quick call just to go through some of the updated features. Great job!

File: PrimerSDK/Classes/User Interface/Primer/PaymentMethodComponent.swift (hunk 7 above)
Comment (EvansPie) [reply PrimerSDK/Classes/User Interface/Primer/PaymentMethodComponent.swift:11]: I was mistaken about this one

File: PrimerSDK/Classes/User Interface/Primer/SingleCardIconComponent.swift (hunk 8 above)
Comment (EvansPie) [reply PrimerSDK/Classes/User Interface/Primer/SingleCardIconComponent.swift]: Removed

File: PrimerSDK/Classes/User Interface/Primer/MultiCardIconComponent.swift (hunk 6 above)
Comment (EvansPie) [reply PrimerSDK/Classes/User Interface/Primer/MultiCardIconComponent.swift]: Removed

File: PrimerSDK/Classes/User Interface/PCI/CardForm/CardFormViewModel.swift (hunk 5 above)
Comment (EvansPie) [reply PrimerSDK/Classes/User Interface/PCI/CardForm/CardFormViewModel.swift]: Removed

**Pull request merged**